pageserver/src/tenant/timeline/init.rs

```diff
@@ -0,0 +1,211 @@
+use crate::{
+    is_temporary,
+    tenant::{
+        ephemeral_file::is_ephemeral_file,
+        remote_timeline_client::{
+            self,
+            index::{IndexPart, LayerFileMetadata},
+        },
+        storage_layer::LayerFileName,
+    },
+    METADATA_FILE_NAME,
+};
+use anyhow::Context;
+use std::{collections::HashMap, ffi::OsString, path::Path, str::FromStr};
+use utils::lsn::Lsn;
+
+/// Identified files in the timeline directory.
+pub(super) enum Discovered {
+    /// The only one we care about
+    Layer(LayerFileName, u64),
+    /// Old ephmeral files from previous launches, should be removed
+    Ephemeral(OsString),
+    /// Old temporary timeline files, unsure what these really are, should be removed
+    Temporary(OsString),
+    /// Temporary on-demand download files, should be removed
+    TemporaryDownload(OsString),
+    /// "metadata" file we persist locally and include in `index_part.json`
+    Metadata,
+    /// Backup file from previously future layers
+    IgnoredBackup,
+    /// Unrecognized, warn about these
+    Unknown(OsString),
+}
+
+/// Scans the timeline directory for interesting files.
+pub(super) fn scan_timeline_dir(path: &Path) -> anyhow::Result<Vec<Discovered>> {
+    let mut ret = Vec::new();
+
+    for direntry in std::fs::read_dir(path)? {
+        let direntry = direntry?;
+        let direntry_path = direntry.path();
+        let file_name = direntry.file_name();
+
+        let discovered = {
+            let fname = file_name.to_string_lossy();
+
+            match LayerFileName::from_str(&fname) {
+                Ok(LayerFileName::Image(file_name)) => {
+                    let file_size = direntry_path.metadata()?.len();
+                    Discovered::Layer(file_name.into(), file_size)
+                }
+                Ok(LayerFileName::Delta(file_name)) => {
+                    let file_size = direntry_path.metadata()?.len();
+                    Discovered::Layer(file_name.into(), file_size)
+                }
+                Err(_) => {
+                    if fname == METADATA_FILE_NAME {
+                        Discovered::Metadata
+                    } else if fname.ends_with(".old") {
+                        // ignore these
+                        Discovered::IgnoredBackup
+                    } else if remote_timeline_client::is_temp_download_file(&direntry_path) {
+                        Discovered::TemporaryDownload(file_name)
+                    } else if is_ephemeral_file(&fname) {
+                        Discovered::Ephemeral(file_name)
+                    } else if is_temporary(&direntry_path) {
+                        Discovered::Temporary(file_name)
+                    } else {
+                        Discovered::Unknown(file_name)
+                    }
+                }
+            }
+        };
+
+        ret.push(discovered);
+    }
+
+    Ok(ret)
+}
+
+/// Decision on what to do with a layer file after considering it's local and remote metadata.
```

Comment:
```suggestion
/// Decision on what to do with a layer file after considering its local and remote metadata.
```

Reply:
Oki, solved it, it was valid future image layer once more.